OutOfSchool/OutOfSchool.Redis/CacheService.cs

```diff
@@ -46,73 +45,92 @@ public async Task<T> GetOrAddAsync<T>(
+        {
+            cacheLock.EnterWriteLock();
             try
             {
-                var value = cache.GetString(key);
-
-                if (value != null)
-                {
-                    returnValue = JsonConvert.DeserializeObject<T>(value);
-                    isExists = true;
-                    return;
-                }
+                cache.Remove(key);
             }
             finally
             {
-                cacheLock.ExitReadLock();
+                cacheLock.ExitWriteLock();
             }
         });
 
-        if (!isExists)
+    public async Task<string?> GetValueFromCacheAsync(string key)
+    {
+        string? returnValue = null;
+        await this.ExecuteRedisMethod(() =>
         {
-            returnValue = await newValueFactory();
-            await ExecuteRedisMethod(() =>
+            cacheLock.EnterReadLock();
+            try
             {
-                cacheLock.EnterWriteLock();
-                try
-                {
-                    var options = new DistributedCacheEntryOptions
-                    {
-                        AbsoluteExpirationRelativeToNow = absoluteExpirationRelativeToNowInterval ?? redisConfig.AbsoluteExpirationRelativeToNowInterval,
-                        SlidingExpiration = slidingExpirationInterval ?? redisConfig.SlidingExpirationInterval,
-                    };
+                returnValue = cache.GetString(key);
 
-                    cache.SetString(key, JsonConvert.SerializeObject(returnValue), options);
-                }
-                finally
+                if (returnValue != null)
                 {
-                    cacheLock.ExitWriteLock();
+                    isExists = true;
                 }
-            });
-        }
+            }
+            finally
+            {
+                cacheLock.ExitReadLock();
+            }
+        });
 
         return returnValue;
     }
 
-    public Task RemoveAsync(string key)
-        => ExecuteRedisMethod(async () => {
+    public async Task SetValueToCacheAsync(string key, string value, TimeSpan? absoluteExpirationRelativeToNowInterval = null, TimeSpan? slidingExpirationInterval = null)
+    {
+        await this.ExecuteRedisMethod(() =>
+        {
             cacheLock.EnterWriteLock();
             try
             {
-                await cache.RemoveAsync(key);
+                var options = new DistributedCacheEntryOptions
+                {
+                    AbsoluteExpirationRelativeToNow = absoluteExpirationRelativeToNowInterval ?? redisConfig.AbsoluteExpirationRelativeToNowInterval,
+                    SlidingExpiration = slidingExpirationInterval ?? redisConfig.SlidingExpirationInterval,
+                };
+
+                cache.SetString(key, value, options);
             }
             finally
             {
                 cacheLock.ExitWriteLock();
             }
         });
+    }
+
+    public Task RemoveFromCacheAsync(string key) => RemoveAsync(key);
```

Comment:
RemoveFromCacheAsync() is removed from CacheService class.
The RemoveAsync() method has been added to ICrudCacheService so that it can be used in MementoService where the ICrudCacheService interface is used.